### PR TITLE
JDK-8088418: Reintroduce JFR Pulse Logger

### PR DESCRIPTION
**Pull request by thegreystone**
Second attempt.

### CODE REVIEW

**Comment by kevinrushforth** [comment]
Tracked in JBS as: [JDK-8088418](https://bugs.openjdk.java.net/browse/JDK-8088418)

**Comment by thegreystone** [comment]
Never contributed to JavaFX before - how on earth do I see what whitespace errors I have? :)

**Comment by kevinrushforth** [comment]
The white-space checker that is run as part of the Travis CI build on Linux and Mac will list the files that are in error, along with the type of error, near the bottom of the log. It won't tell you which lines have the problem. Here is the output from the above Travis CI run on Linux:

```
--- List of added or modified files

modules/javafx.base/src/main/java/com/sun/javafx/logging/PrintLogger.java
modules/javafx.base/src/main/java/com/sun/javafx/logging/PulseLogger.java
modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRInputEvent.java
modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseEvent.java
modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseLogger.java
modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/PulseId.java
modules/javafx.base/src/main/java/module-info.java

--- Checking for white space errors

modules/javafx.base/src/main/java/com/sun/javafx/logging/PulseLogger.java :trailingWhitespace:
modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseLogger.java :trailingWhitespace:

Found 2 whitespace or executable issues

To correct, use
   bash tools/scripts/checkWhiteSpace -F -x

Error in white space check
```

To fix them, you can either do it manually (e.g., using your IDE to remove trailing white-spaces), or you can run:

```
cat files.list | bash tools/scripts/checkWhiteSpace -S -F -x
```

where `files.list` has the names of files you want to process (the two file names with the issue)

**Comment by thegreystone** [comment]
Seems all is fine now. :)

File: modules/javafx.base/src/main/java/com/sun/javafx/logging/PrintLogger.java [hunk 1]
```diff
@@ -394,7 +385,7 @@ void printAndReset() {
                 System.err.print(message);
                 if (!counters.isEmpty()) {
                     System.err.println("Counters:");
-                    List<Map.Entry<String,Counter>> entries = new ArrayList(counters.entrySet());
+                    List<Map.Entry<String,Counter>> entries = new ArrayList<>(counters.entrySet());
                     Collections.sort(entries, (a, b) -> a.getKey().compareTo(b.getKey()));
                     for (Map.Entry<String, Counter> entry : entries) {
                         System.err.println("\t" + entry.getKey() + ": " + entry.getValue().value);
```
[the remark this answers is not in the document]
Comment (thegreystone) [reply modules/javafx.base/src/main/java/com/sun/javafx/logging/PrintLogger.java]: I opted not to change pre-existing code.

File: modules/javafx.base/src/main/java/com/sun/javafx/logging/PrintLogger.java [hunk 2]
```diff
@@ -355,7 +346,7 @@ public void newPhase(String name) {
         int pulseCount;
         boolean pushedRender;
         StringBuffer message = new StringBuffer();
-        Map<String,Counter> counters = new ConcurrentHashMap();
+        Map<String,Counter> counters = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (thegreystone) [reply modules/javafx.base/src/main/java/com/sun/javafx/logging/PrintLogger.java]: I opted not to change pre-existing code. I just got rid of the compiler warning.

**Comment by thegreystone** [comment]
Kevin, is there anything else you need from me?

**Comment by kevinrushforth** [comment]
No, I just need make some time to review it.

**Comment by thegreystone** [comment]
Let me know. :)

**Comment by thegreystone** [comment]
Haha. No worries. Busy times here too. :)

File: modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRInputEvent.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2014, Oracle and/or its affiliates. All rights reserved.
```
Comment (kevinrushforth) [inline modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRInputEvent.java]: Should be either `2018,` or `2014,2018,`

File: modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseEvent.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2014, Oracle and/or its affiliates. All rights reserved.
```
Comment (kevinrushforth) [inline modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseEvent.java]: Should be either `2018,` or `2014,2018,`

File: modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseLogger.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2014, Oracle and/or its affiliates. All rights reserved.
```
Comment (kevinrushforth) [inline modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseLogger.java]: Should be either `2018,` or `2014,2018,`

File: modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/PulseId.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package com.sun.javafx.logging.jfr;
```
Comment (kevinrushforth) [inline modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/PulseId.java]: This file needs a proper Oracle GPLv2+Classpath header.

File: modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseLogger.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+    private int pulseNumber;
+    private int fxPulseNumber;
+    private int renderPulseNumber;
+    private Thread fxThread;
+
+    public static Logger createInstance() {
+        // Since events are off by default, and overhead is low, we can always
+        // have this pulse logger available.
+        return new JFRPulseLogger();
```
[the remark this answers is not in the document]
Comment (thegreystone) [reply modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseLogger.java]: This, however, means the JFR pulse logger will only work if JFR was enabled on startup - i.e. you cannot enable the JFR pulse logger dynamically. Perhaps this is an okay constraint?

[the remark this answers is not in the document]
Comment (kevinrushforth) [reply modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseLogger.java]: I think this is best for now, and matches what was done in FX 8. We could consider an RFE to allow dynamic enabling of this in the future if there is a need, but I think this constraint is fine.

[the remark this answers is not in the document]
Comment (thegreystone) [reply modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseLogger.java]: How about if flight recorder is initialised, or javafx.pulseLogger is set? If someone has JFR enabled, I think it's fair to assume they want the events.

[the remark this answers is not in the document]
Comment (thegreystone) [reply modules/javafx.base/src/main/java/com/sun/javafx/logging/jfr/JFRPulseLogger.java]: Committed a version doing this. Let me know what you think.

**Comment by thegreystone** [comment]
Meanwhile, I’ll take a look at making the JavaFX page in JMC support both the old JavaFX 8 events and the new ones. 

See: https://bugs.openjdk.java.net/browse/JMC-6277

**Comment by egahlin** [comment]
I think better names could be picked for the events. Today they lack @Name which means JFR will default to the package name "com.sun.javafx.logging.jfr". 

The event name is used to identify the event for an applicatio consuming the events (using a parser). It is also used to configure event settings (programmatically or using a configuration file.)  It is important to get the event names correct, because changing them will break any use of it. @Label, @Description and @Category can on the otherhand can be changed more freely..

Not sure what would constitute a good name, but perhaps "openjdk.jfx.Pulse" and "openjdk.jfx.Input"? There is no need to include "jfr" or "event" as this is implicit.  "Event" may be OK if the event is actually tracking another type of event, like an AWTEvent.

The name of the relationad id could be improved, perhaps "openjdk.jfx.PulseId"?

**Comment by kevinrushforth** [comment]
> I think better names could be picked for the events. Today they lack @name which means JFR will default to the package name "com.sun.javafx.logging.jfr".

Thank you for raising this issues.

The deadline for getting this in for openjfx 12 is Monday (unless we make an exception for getting an enhancement in after RDP1). This may need more discussion than we have time for prior to that. Depending on what @thegreystone wants to do we could go ahead and get this in for RDP1 with the existing names, and then file a follow-on bug to fix the names, although if we follow that approach, we would need to fix the names in openjfx12 (before the end of January) to avoid the compatibility problems you mentioned above.

> Not sure what would constitute a good name, but perhaps "openjdk.jfx.Pulse" and "openjdk.jfx.Input"?

Is there an existing pattern of prefixing these events with "openjdk"? Otherwise, I would probably suggest "javafx.jmc" or "jmc.javafx" or as a prefix. Or, since this is in a name space where you already know that it's a JMC event, maybe just "javafx". Are there similar events for other components in the JDK?

**Comment by egahlin** [comment]
Reversed domain name notation is what I would recommend in general.

That said, JavaFX is not the typical product and I noticed "javafx.version" was used as system property. Nobody building a JavaFX application would prefix their system properties or JFR events with "javafx." so I think that would work out fine. FWIW, the events in the JDK is prefixed with "jdk."

I don't think we should put in "jmc" as this is the name of the client .Sounds fine to check in what exists today and change the names later as a follow up issue.

**Comment by thegreystone** [comment]
If we want to prettify the names, I vote ”javafx.”.

**Comment by kevinrushforth** [comment]
The above change from Pulse --> PulsePhase looks good.

**Pull request merged**